### PR TITLE
fix: handle destroyed buildings properly

### PR DESCRIPTION
**Pull request by winex**
**READY FOR TESTING**
**DEPENDS:** ~~#307, #308~~ _don't squash, pls_

because `on_entity_destroyed` event doesn't provide `entity.position` (it's already gone!) we need to remember `.position` field in `global.tycoon_city_buildings[]` and free this cell by new `City::freeCellAtPosition()` function.
_`City::clearCell()` does different thing, sadly_ - so i'm reworking it...

### notes
i've fixed adding **all** entities from `Constants.CITY_SPECIAL_BUILDINGS` to `global.tycoon_city_buildings[]`, since code was already doing it, but not fully. this one probably should have been named as `global.tycoon_buildings`
while this could be split into _city-built_ and _player-built_ lists, it would require checking two arrays in event handlers.

ok, this was harder than i thought before. sorry for such big changes, but i couldn't find another way. diffs are horrible, looking at resulting files is easier
my tests was successful, i've tried to avoid `assert()`s so any save should still be playable, but couldn't decide in a few places (iirc). will fix asap if you find anything!

### tl;dr
- new function `City::freeCellAtPosition()`
- new functions `Util:: {add,remove,get}GlobalBuilding()`
- `global.tycoon_city_buildings[]` must now have every related `tycoon-*` entity, or `on_removed()` handler will skip it
- we always must `script.register_on_entity_destroyed(entity)` after creation
- migration script should fix possible errors in old saves
- logging might be a bit excessive, but helps testing. we can reduce it later
- **NOTE:** if any market (hw-store, etc...) is missing for a city after migration resets special buildings cache - it needs to be rebuilt or replaced closer
- _using `CITY_RADIUS` differs a little when placing vs `find_entities_*()` probably because `town-hall` vs `city.center`_

### ~~todo~~ migration
- [x] finish position processing correctly
- [x] process each city grid looking for buildings/etc, as it could be wrong
- [x] add missing entities ~~to `global.tycoon_city_buildings[]`~~ by `Util.addGlobalBuilding()`

nothing else would be needed at runtime, as new buildings will already work

### CODE REVIEW

File: city.lua [hunk 1]
```diff
@@ -1325,6 +1326,24 @@ local function clearCell(city, upgradeCell)
...
+    end
+    -- check for incorrect cell
+    assert(cell.type == "building")
+
+    local p = cellCoordinates
+    city.grid[p.y][p.x] = {type = "unused"}
+    -- this doesn't include actual position, so shift to the top-left
+    addBuildingLocations(city, {x = p.x - 1, y = p.y - 1})
+    -- TODO: can't we just add it like this?
+    --city.buildingLocationQueue.pushleft(p)
```
Comment (winex) [inline city.lua:1344]: `addBuildingLocations()` works nicely here.
i'd prefer just to push this location into the building queue, but couldn't understand these code paths fully.

maybe i need to try some call visualizer -- still don't get it after reading files so many times :(

Comment (bahrmichael) [reply city.lua:1344]: Did you try this?

```
Queue.pushleft(city.buildingLocationQueue, coordinates)
```

Unfortunately the methods from `queue.lua` are only available via the `Queue` call, not on the field that we see as a queue.

See this as an example:

https://github.com/bahrmichael/factorio-tycoon/blob/49c2789f9e55dd1b441b868c45d08d4e4d3f4880/city.lua#L699

**Comment by bahrmichael** [comment]
I like this! I left the code to restore houses in a bit of a broken state unfortunately. Thank you for picking this up!

File: city.lua (hunk 1 above)
Comment (winex) [reply city.lua]: i've tried this, but it **just forces** to build in that position. so, when it is called somewhere inbetween house upgrade - it just overlaps with second entity in that cell :(
also, i can't find proper place to add a check if that position in `city.buildingLocationQueue` is already occupied (for any reason)

ofc, i moved `script.register_on_entity_destroyed(entity)` to register most of our city entities (like gardens), otherwise nothing will be able to clear cells at runtime

**Comment by winex** [comment]
i have a question so i won't revert your refactoring w/o reason:
- constructors and destructors (in C++ terms) must be in the same file close together, so we never miss things
- `construction-event-handler` currently does some citizen recalc and lights removal from `global`

is it fine to move this part back into `City::` and sync it? and it probably should only callback related methods, without actually checking any internals, smth like this pseudo-code:
```
local function on_some_event(event)
    if this_is_related_to_city then
        City.destroyHouse(city, event.bulding.posGrid)
    end
    if this_is_related_to_something_else then
        something_else.callback(event.entity_name, ...)
    end
end
```

**Comment by bahrmichael** [comment]
> is it fine to move this part back into City:: and sync it?

I'm fine with that! For the most of it, don't hold strong opinions on lua coding because I lack the experience with it.

File: city.lua (hunk 1 above)
Comment (bahrmichael) [reply city.lua]: > i've tried this, but it just forces to build in that position. so, when it is called somewhere inbetween house upgrade - it just overlaps with second entity in that cell :(
also, i can't find proper place to add a check if that position in city.buildingLocationQueue is already occupied (for any reason)

I think I just forced so far, and didn't build safety checks down the line.

When you add an entry to a location queue, the code should eventually run through city#startConstruction. A few lines in there is a piece of code that removes duplicates from the queue. Iirc this was a bandaid that worked well enough to make it to production :D

https://github.com/bahrmichael/factorio-tycoon/blob/4579b5de58fa6cb04374a20d2a50cc914a132813/city.lua#L829-L831

To check if a location is already occupied, I would check the city's grid, translate the grid to map position and then search for entities there (if the city's grid has not been updated with an on_destroy hook).

I hope that helps, let me know if you have more questions!

Comment (winex) [reply city.lua]: oh, that's why! i was searching for `buildingLocationQueue`, but here it is
```
coordinates = Queue.popleft(city[queueIndex])
```
with `queueIndex = "buildingLocationQueue"`, missed that. thanks!

**Comment by winex** [comment]
_after reworking destruction/clearing code paths..._
oh, lol! totally missed the point that `city.grid` is being enlarged over time from `{1, 1}`, so my `.posGrid` becomes definitely wrong. if it was centered and supported negative coordinates - it would work perfectly...

reworking `city.grid` would require very accurate migration script - i'm not ready for this, yet. i'm writing like 4th 
migration already, which is not good...

then, we must remember every building's `.position` instead :(
_sometimes Factorio makes simple things sooo hard and ugly_

**Comment by winex** [comment]
```lua
-- event.name:
-- on_player_entity_mined = 67
-- on_entity_destroyed = 160
```
ok, this is why we must remember `.position` for every building:
```lua
-- valid: true
75070.809 Script @__tycoon__/construction-event-handler.lua:103: on_removed(): event: 67 unit: 394 valid: true position: {x = 137, y = 277} city: Ironhaven name: tycoon-house-simple-6
-- valid: false
75131.876 Script @__tycoon__/construction-event-handler.lua:103: on_removed(): event: 160 unit: 6499 valid: false position: {x = 179, y = 265} city: Ironhaven name: tycoon-house-simple-8
```
mined first one, but the last one i machine-gunned...

`garden` mining works (and city rebuilt it), but destruction differs - do we put it into `global.tycoon_city_buildings` as well?

**Comment by winex** [comment]
this possible mismatch hell (separate commit, just to emphasize) took last pieces of my energy today!
just can't handle every possible thing in the universe... probably nobody will understand this code fully :(

good thing that it works nicely now, except that i couldn't catch event for supply buildings and train-stations even if i register them - idk what's wrong there

i'll try to split it a bit, so the change is not that big: new functions + change.
then finish city grid parsing when i get what to do there. i have few saves that could be good test case

but now i need to sleep, lol...

**Comment by winex** [comment]
this town was rebuilt succesfully few times (>10 buildings) around town-hall:
scrot 20240402_050058 city-rebuilding crop half

**Comment by winex** [comment]
the last thing depends on mentioned PR, i'll rebase this after that one is merged

fixed and resolved few more things, it looks way better now

**Comment by winex** [comment]
one last step (from todo) and this is going to be ready for testing...

**Comment by winex** [comment]
when machine-gunning `tycoon-treasury`, city can't rebuild it, but this is another issue...

**Comment by winex** [comment]
ok, marking this as ready for testing! i've been working on top of this branch already and everything worked fine.

i'll rebase one last time after deps are fully merged in...

File: migrations/0.4.6-fix-city-buildings.lua [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    -- drop cache
+    for name, _ in pairs(Constants.CITY_SPECIAL_BUILDINGS) do
+        city.special_buildings.other[name] = nil
+    end
+
+    -- check city grid
+    for y = 1, #city.grid, 1 do
+        for x = 1, #city.grid, 1 do
+            local row = city.grid[y]
+            local cell = ((row or {})[x] or {}).cell
```
Comment (winex) [inline migrations/0.4.6-fix-city-buildings.lua:99]: oh, this is wrong!

File: migrations/0.4.6-fix-city-buildings.lua [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            if cell == nil or cell.type ~= "building" then
+                goto continue
+            end
+
+            -- if building is absent
+            local entity = cell.entity
+            if entity == nil or (not entity.valid) then
+                City.freeCellAtPosition(city, {x=x, y=y})
```
Comment (winex) [inline migrations/0.4.6-fix-city-buildings.lua:107]: and this function accepts `MapPosition`, not `GridPosition`. fixing...

File: migrations/0.4.6-fix-city-buildings.lua [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                    cell.entity = city.special_buildings.town_hall
+                    log("assigned town hall to cell.entity!")
+                    goto continue
+                elseif cell.initKey ~= nil then
+                    log(string.format("unknown initKey, cell: ", serpent.line(cell)))
+                end
+
+                -- clear manually
+                city.grid[y][x] = { type = "unused" }
+                City.updatepossibleBuildingLocations(city, cellCoordinates, true)
```
Comment (winex) [inline migrations/0.4.6-fix-city-buildings.lua:118]: oops

**Comment by winex** [comment]
fixed my own errors and it looks like i've tested almost every code path now. this work was hard and i'm glad it's nearing to an end.
migration works nicely and fixes many saves with different states as it can be seen in the logs

@bahrmichael waiting for feedback of your tests...

**Comment by winex** [comment]
rebased on updated #308 (avoid `goto`), reused utility functions

`global.tycoon_city_buildings` now only exists in `util.lua` wrappers and migrations (as latter probably should depend on minimal amount of utility calls)

**Comment by winex** [comment]
github was a bit wrong in conflicts again, now it's cleaner than ever! ;)

File: construction-event-handler.lua
Comment (winex) [inline construction-event-handler.lua:1]: this resulted in `on_removed()` function be simplified to a minimum... city-related checks (of internal state) are now back in `city.lua`

File: construction-event-handler.lua [hunk 5]
```diff
@@ -30,12 +31,17 @@ local function on_built(event)
 
         invalidateSpecialBuildingsList(city, entity.name)
 
+        -- TODO: this could be dropped in favor of Util.getGlobalBuilding()
         if global.tycoon_entity_meta_info == nil then
             global.tycoon_entity_meta_info = {}
         end
         global.tycoon_entity_meta_info[entity.unit_number] = {
             cityId = city.id
         }
```
Comment (winex) [inline construction-event-handler.lua:40]: `global.tycoon_entity_meta_info` is now being a (subset-) duplicate of `global.tycoon_city_buildings`, but i've just marked it with TODO to keep this PR as small as possible to be tested and merged asap

File: city.lua [hunk 6]
```diff
@@ -721,6 +721,8 @@ local function addBuildingLocations(city, recentCoordinates, allowCenter)
     if recentCoordinates ~= nil then
         local surrounds = getSurroundingCoordinates(recentCoordinates, 1, false, allowCenter)
         for _, value in ipairs(surrounds) do
+            -- TODO: this will probably skip corners when called from on_entity_destroyed event
+            -- are edges supposed to be roads or what?
```
Comment (bahrmichael) [inline city.lua:725]: The reason for this check lies within the self-expanding city grid. During road expansion we may detect that we've reached the edge of the grid: https://github.com/bahrmichael/factorio-tycoon/blob/888831dee6f2f586b82de6013882f281a5f40fab/city.lua#L972-L986

When that happens, it adds a new row/column at each side of the grid to give more space to grow into.

The check in this file just says "if the grid hasn't gone far enough yet, then ignore that section and try again in the future". This leads to the city having a higher chance to fill out the whole grid, instead of growing along one tree.

File: city.lua [hunk 7]
```diff
@@ -1390,25 +1405,130 @@ local function findUpgradableCells(city, limit, upgradeTo)
...
+    --log("removeBuilding(): building: ".. serpent.line(building))
+
+    --- NOTE: switch city reference block, common to all the checks below
+    city = city or Util.findCityById(building.cityId)
+    if city == nil then
+        log("removeBuilding(): ERROR: unable to get city reference!")
+    end
+    -- WARN: this weird case should never happen, but we must always check
+    if city ~= nil and city.id ~= building.cityId then
+        local msg = string.format("removeBuilding(): ERROR: city.id: %s doesn't match building.cityId: %s, unit_number: %d",
```
Comment (bahrmichael) [inline city.lua:1431]: Let's prefix this with "Tycoon" so players know which mod this message is coming from.

File: util.lua [hunk 8]
```diff
@@ -274,6 +275,66 @@ local function findCityByEntityUnitNumber(unitNumber)
...
+local function removeGlobalBuilding(unit_number)
+    if global.tycoon_city_buildings == nil or unit_number == nil then
+        return
+    end
+
+    global.tycoon_city_buildings[unit_number] = nil
+end
+
+--- @param unit_number number
+--- @return Building
```
Comment (bahrmichael) [inline util.lua:328]: This can also return nil.

File: util.lua [hunk 9]
```diff
@@ -274,6 +275,66 @@ local function findCityByEntityUnitNumber(unitNumber)
...
+--- @field cityId number
+--- @field entity LuaEntity
+--- @field entity_name string
+--- @field position MapPosition | Coordinates
+--- @field isSpecial boolean | nil
+
+--- @param unit_number number
+--- @param cityId number
+--- @param entity LuaEntity | nil
+--- @return Building
```
Comment (bahrmichael) [inline util.lua:288]: This can also return nil.

**Comment by bahrmichael** [comment]
I'll test and merge this later today 👍 Thank you very much for your deep dive into the structure handling!

**Comment by winex** [comment]
_sorry, missed window for review changes and can't add anything more after this was merged and closed, so i created another little PR_

> Just tested by nuking a town, and eventually it rebuilt. It took much longer than expected (building houses in many other places first), but eventually got back to the destroyed houses.

yeah, when `buildingLocationQueue` gets bigger and bigger, like 30-ish or more, it takes time to get to "new entries". i used `addBuildingLocations()` because it does many more checks, so nothing goes wrong for now.

it probably should be changed somehow to iterate from city center in circles as i commented here _(i think L1529 can still happen if force-pushed into queue)_:
https://github.com/bahrmichael/factorio-tycoon/blob/8956c35c5e4119daaad2f95dfdabf248929afefa/city.lua#L1527-L1531
didn't try to change it (back then) to keep this important PR minimal. thanks for accepting my contributions! :+1:
